bridge/tracker/tracker.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	return nil
+}
+
+//	startEventTracking handles the subscription object (provided by the rootchain client).
+func (t *Tracker) startEventTracking() {
+	for {
+		select {
+		//	process new header
+		case header := <-t.sub.newHeadCh():
+			t.trackHeader(header)
+
+		//	handle sub error
+		case err := <-t.sub.errCh():
+			t.ctxSubscription.cancel()
+			t.logger.Error("subscription cancelled", err)
+
+			return
+
+		//	stop tracker's sub
+		case <-t.ctxSubscription.done():
+			if err := t.sub.unsubscribe(); err != nil {
+				t.logger.Error("cannot unsubscribe", "errCh", err)
+
+				return
+			}
+
+			t.logger.Debug("subscription stopped")
+
+			return
+		}
+	}
+}
+
+//	trackHeader determines the range of block to query
+//	based on the blockNumber of the header and issues an appropriate
+//	eth_getLogs call. If an event of interest was emitted,
+//	it is sent to eventCh.
+func (t *Tracker) trackHeader(header *ethHeader) {
+	//	determine range of blocks to query
+	fromBlock, toBlock := t.calculateRange(header)
+	if fromBlock == nil || toBlock == nil {
+		//	we are returning here because the calculated
+		//	range was already queried or the chain is not
+		//	at the desired depth.
+		return
+	}
+
+	t.logger.Info("querying events within block range", "from", fromBlock, "to", toBlock)
+
+	//	fetch logs
+	logs := t.queryEvents(fromBlock, toBlock)
+	t.logger.Info("matched events", "num", len(logs))
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to *big.Int) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
+	latestHeight := big.NewInt(0).SetUint64(header.Number)
+
+	//	or: TODO ethereum header
+	//latestHeight, _ := types.ParseUint256orHex(&header.Number)
+
+	//	check if block number is at required depth
+	if latestHeight.Cmp(t.confirmations) < 0 {
+		//	block height less than required
+		t.logger.Debug(
+			"not enough confirmations",
+			"current", latestHeight.Uint64(),
+			"required", t.confirmations.Uint64())
+
+		return
+	}
+
+	//	right bound
+	toBlock := big.NewInt(0).Sub(latestHeight, t.confirmations)
+
+	//	left bound
+	fromBlock := t.loadLastBlock()
+	if fromBlock == nil {
+		fromBlock = toBlock
+	} else {
+		//	increment - last block number was already queried
+		fromBlock = fromBlock.Add(fromBlock, big.NewInt(1))
+	}
+
+	if toBlock.Cmp(fromBlock) < 0 {
+		//	no need to perform any query
+		//	as left bound is higher than the right
+		return
+	}
+
+	return fromBlock, toBlock
+}
+
+//	loadLastBlock returns the block number of the last
+// 	block processed by the tracker (if available).
+func (t *Tracker) loadLastBlock() *big.Int {
+	lastBlockNumber, ok := t.db.ReadHeadNumber()
+	if !ok {
+		return nil
+	}
+
+	return big.NewInt(0).SetUint64(lastBlockNumber)
+}
+
+//	saveLastBlock stores the number of the last block processed
+//	into the tracker's database.	.
```

Comment:
Nitpick: all method comments should be without the punctuation at the end.

Please make sure we keep the code guidelines in check regarding this detail 🙏

Reply:
[Fixed](https://github.com/0xPolygon/polygon-edge/pull/386/commits/c3b941e0d61fa63d21c363c9e89a23807e507e35)